commit 9fd6e74a774660f1ef2dfdc32ebe7cc875c1ef86
Author: ejc3 <ejc3@users.noreply.github.com>
Date:   Sun Jan 11 00:21:18 2026 +0000

    virtio: Flush vring cache before kick for ARM64 NV2
    
    Under ARM64 nested virtualization (FEAT_NV2), the hypervisor may read
    stale data from virtqueue ring structures unless explicit cache
    maintenance is performed.
    
    Signed-off-by: fcvm <fcvm@localhost>

diff --git a/drivers/virtio/virtio_ring.c b/drivers/virtio/virtio_ring.c
index 7b6205253b46..53b388ae9feb 100644
--- a/drivers/virtio/virtio_ring.c
+++ b/drivers/virtio/virtio_ring.c
@@ -15,6 +15,11 @@
 #include <linux/spinlock.h>
 #include <xen/xen.h>
 
+#ifdef CONFIG_ARM64
+#include <asm/barrier.h>
+#include <asm/cacheflush.h>
+#endif
+
 #ifdef DEBUG
 /* For development, we want to crash whenever the ring is screwed. */
 #define BAD_RING(_vq, fmt, args...)				\
@@ -2489,6 +2494,36 @@ bool virtqueue_notify(struct virtqueue *_vq)
 	if (unlikely(vq->broken))
 		return false;
 
+#ifdef CONFIG_ARM64
+	/*
+	 * NV2 cache coherency: Flush all vring structures to ensure
+	 * hypervisor sees updated descriptors and available ring.
+	 * DSB alone does not flush dirty cache lines under nested virt.
+	 */
+	dsb(sy);
+	if (!vq->packed_ring) {
+		/* Split virtqueue: flush desc, avail, and used rings */
+		struct vring *vr = &vq->split.vring;
+		unsigned long desc_start = (unsigned long)vr->desc;
+		unsigned long desc_end = desc_start + (16 * vr->num);
+		unsigned long avail_start = (unsigned long)vr->avail;
+		unsigned long avail_end = avail_start +
+			sizeof(struct vring_avail) + (2 * vr->num);
+		unsigned long used_start = (unsigned long)vr->used;
+		unsigned long used_end = used_start +
+			sizeof(struct vring_used) + (8 * vr->num);
+
+		dcache_clean_inval_poc(desc_start, desc_end);
+		dcache_clean_inval_poc(avail_start, avail_end);
+		dcache_clean_inval_poc(used_start, used_end);
+	} else {
+		/* Packed virtqueue: TODO if needed */
+		dsb(sy);
+	}
+	dsb(sy);
+	isb();
+#endif
+
 	/* Prod other side to tell it about changes. */
 	if (!vq->notify(_vq)) {
 		vq->broken = true;
